### PR TITLE
Add staking module README

### PR DESCRIPTION
**Pull request by Raneet10**
# Description

This PR adds README for `staking` module and fixes a few typos. 

# Changes

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [ ] Changes only for a subset of nodes

# Checklist

- [x] I have added at least 2 reviewer or the whole pos-v1 team
- [x] I have added sufficient documentation in code
- [x] I will be resolving comments - if any - by pushing each fix in a separate commit and linking the commit hash in the comment reply

## Testing

- [ ] I have added unit tests
- [ ] I have added tests to CI
- [ ] I have tested this code manually on local environment
- [ ] I have tested this code manually on remote devnet using express-cli
- [ ] I have tested this code manually on mumbai
- [ ] I have created new e2e tests into express-cli

### CODE REVIEW

File: staking/README.md [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+* [How to propose a MsgValidatorExit transaction](#how-to-propose-a-msgvalidatorexit-transaction)
+* [How does a validator update its stake](#how-does-a-validator-update-its-stake)
+* [How to propose a MsgStakeUpdate transaction](#how-to-propose-a-msgstakeupdate-transaction)
+* [How does a validator update its signer address](#how-does-a-validator-update-its-signer-address)
+* [How to propose a MsgSignerUpdate transaction](#how-to-propose-a-msgsignerupdate-transaction)
+* [Query commands](#query-commands)
+
+## Preliminary terminology
+
+* An `epoch` represents the period till a checkpoint is submitted on Ethereum, i.e one `epoch` ends when a checkpoint is committed on Ethereum and the next one begins.
```
Comment (marcello33) [inline staking/README.md:19]: i.e.

Comment (marcello33) [inline staking/README.md:19]: Not sure, but I'd probably say "An `epoch` represents the period in which a checkpoint is being submitted on Ethereum (i.e. one `epoch` ends when the checkpoint is committed on and the next one begins.)"

File: staking/README.md [hunk 2]
```diff
@@ -0,0 +1,541 @@
+# Staking Module
+
+## Table of Contents
+
+* [Preliminary terminology](#preliminary-terminology)
+* [Overview](#overview)
+* [How does one join the network as a validator](#how-does-one-join-the-network-as-a-validator)
+* [How to propose a MsgValidatorJoin transaction](#how-to-propose-a-msgvalidatorjoin-transaction)
```
Comment (marcello33) [inline staking/README.md:8]: Don't you think the `propose` sections are sub-paragraphs of the previous one?
e.g.
```
* [How does one join the network as a validator](#how-does-one-join-the-network-as-a-validator)
** [How to propose a MsgValidatorJoin transaction](#how-to-propose-a-msgvalidatorjoin-transaction)
```

File: staking/README.md [hunk 3]
```diff
@@ -0,0 +1,541 @@
...
+The `staking` module in Heimdall is responsible for a validator's stake related operations. It primarily aids in
+
+* A node joining the protocol as a validator.
+* An node leaving the protocol as a validator.
+* Updating an existing validator's stake in the network.
+* Updating the signer address of an existing validator.
+
+## How does one join the network as a validator
+
+The node that wants to be a validator stakes its tokens by invoking the `stakeFor` method on the `StakeManager` contract on L1(Ethereum), which emits a `Staked` event:
```
Comment (marcello33) [inline staking/README.md:32]: L1 (Ethereum)

File: staking/README.md [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+	ActivationEpoch uint64                  `json:"activationEpoch"`
+	Amount          sdk.Int                 `json:"amount"`
+	SignerPubKey    hmTypes.PubKey          `json:"pub_key"`
+	TxHash          hmTypes.HeimdallHash    `json:"tx_hash"`
+	LogIndex        uint64                  `json:"log_index"`
+	BlockNumber     uint64                  `json:"block_number"`
+	Nonce           uint64                  `json:"nonce"`
+}
+```
+where ,
```
Comment (marcello33) [inline staking/README.md:68]: I'd remove the comma

File: staking/README.md [hunk 5]
```diff
@@ -0,0 +1,541 @@
...
+
+* `From` represents the address of the validator that initiated the `MsgValidatorJoin` transaction on heimdall.
+* `ID` represents the id of the new validator.
+* `ActivationEpoch` is the `epoch` at which the new validator will be activated.
+* `Amount` is the total staked amount.
+* `SignerPubKey` is the signer public key of the new validator.
+* `TxHash` is the hash of the staking transaction on L1.
+* `LogIndex` is the index of the `Staked` log in the staking transaction receipt.
+* `BlockNumber` is the L1 block number in which the staking transaction was included.
+* `Nonce` is the the count representing all the staking related transactions performed from the new validator's account. This is to meant keep Heimdall and L1 in sync.
```
Comment (marcello33) [inline staking/README.md:78]: "This is meant to keep"

File: staking/README.md [hunk 6]
```diff
@@ -0,0 +1,541 @@
...
+
+Upon broadcasting the message, it goes through `HandleMsgValidatorJoin` handler which checks the basic sanity of the transaction (verifying the validator isn't already existing, voting power, etc.).
+
+The `SideHandleMsgValidatorJoin` side-handler in all the existing (honest) validators then ensures the authenticity of staking transaction on L1. It fetches transaction receipt from L1 contract and validates it with the data provided in the `MsgValidatorJoin` transaction. Upon successful validation, `YES` is voted.
```
Comment (marcello33) [inline staking/README.md:82]: it fetches the transaction receipt

File: staking/README.md [hunk 7]
```diff
@@ -0,0 +1,541 @@
...
+	From              hmTypes.HeimdallAddress `json:"from"`
+	ID                hmTypes.ValidatorID     `json:"id"`
+	DeactivationEpoch uint64                  `json:"deactivationEpoch"`
+	TxHash            hmTypes.HeimdallHash    `json:"tx_hash"`
+	LogIndex          uint64                  `json:"log_index"`
+	BlockNumber       uint64                  `json:"block_number"`
+	Nonce             uint64                  `json:"nonce"`
+}
+```
+where ,
```
Comment (marcello33) [inline staking/README.md:194]: i'd remove the comma

File: staking/README.md [hunk 8]
```diff
@@ -0,0 +1,541 @@
...
+	From        hmTypes.HeimdallAddress `json:"from"`
+	ID          hmTypes.ValidatorID     `json:"id"`
+	NewAmount   sdk.Int                 `json:"amount"`
+	TxHash      hmTypes.HeimdallHash    `json:"tx_hash"`
+	LogIndex    uint64                  `json:"log_index"`
+	BlockNumber uint64                  `json:"block_number"`
+	Nonce       uint64                  `json:"nonce"`
+}
+```
+where ,
```
Comment (marcello33) [inline staking/README.md:277]: i'd remove the comma

File: staking/README.md [hunk 9]
```diff
@@ -0,0 +1,541 @@
...
+	From            hmTypes.HeimdallAddress `json:"from"`
+	ID              hmTypes.ValidatorID     `json:"id"`
+	NewSignerPubKey hmTypes.PubKey          `json:"pubKey"`
+	TxHash          hmTypes.HeimdallHash    `json:"tx_hash"`
+	LogIndex        uint64                  `json:"log_index"`
+	BlockNumber     uint64                  `json:"block_number"`
+	Nonce           uint64                  `json:"nonce"`
+}
+```
+where ,
```
Comment (marcello33) [inline staking/README.md:358]: i'd remove the comma

File: staking/README.md (hunk 2 above)
Comment (Raneet10) [reply staking/README.md:8]: Good idea, though I just followed the format from `bor` module.

File: staking/README.md (hunk 1 above)
Comment (Raneet10) [reply staking/README.md:19]: Not sure whether you're pointing to the wording or suggesting the latter part be added in brackets or both but how about `An epoch represents the period until a checkpoint is submitted on Ethereum (i.e. one `epoch` ends when a checkpoint is committed on Ethereum and the next one begins).` ?

Comment (Raneet10) [reply staking/README.md]: Rectified: [159a33b](https://github.com/maticnetwork/heimdall/pull/1082/commits/159a33bafa2f43c7cc2ae0f9d57b4257ef036436)

Comment (Raneet10) [reply staking/README.md]: See [159a33b](https://github.com/maticnetwork/heimdall/pull/1082/commits/159a33bafa2f43c7cc2ae0f9d57b4257ef036436)

File: staking/README.md (hunk 2 above)
Comment (Raneet10) [reply staking/README.md:8]: Check how this looks: [159a33b](https://github.com/maticnetwork/heimdall/pull/1082/commits/159a33bafa2f43c7cc2ae0f9d57b4257ef036436)

File: staking/README.md (hunk 3 above)
Comment (Raneet10) [reply staking/README.md]: Rectified: [159a33b](https://github.com/maticnetwork/heimdall/pull/1082/commits/159a33bafa2f43c7cc2ae0f9d57b4257ef036436)

File: staking/README.md (hunk 4 above)
Comment (Raneet10) [reply staking/README.md]: So did I: [159a33b](https://github.com/maticnetwork/heimdall/pull/1082/commits/159a33bafa2f43c7cc2ae0f9d57b4257ef036436)

File: staking/README.md (hunk 5 above)
Comment (Raneet10) [reply staking/README.md]: Fixed: [159a33b](https://github.com/maticnetwork/heimdall/pull/1082/commits/159a33bafa2f43c7cc2ae0f9d57b4257ef036436)

File: staking/README.md (hunk 6 above)
Comment (Raneet10) [reply staking/README.md]: Recitifed: [159a33b](https://github.com/maticnetwork/heimdall/pull/1082/commits/159a33bafa2f43c7cc2ae0f9d57b4257ef036436)

File: staking/README.md (hunk 7 above)
Comment (Raneet10) [reply staking/README.md]: Done: [159a33b](https://github.com/maticnetwork/heimdall/pull/1082/commits/159a33bafa2f43c7cc2ae0f9d57b4257ef036436)

File: staking/README.md (hunk 8 above)
Comment (Raneet10) [reply staking/README.md]: Done: [159a33b](https://github.com/maticnetwork/heimdall/pull/1082/commits/159a33bafa2f43c7cc2ae0f9d57b4257ef036436)

File: staking/README.md (hunk 9 above)
Comment (Raneet10) [reply staking/README.md]: Done: [159a33b](https://github.com/maticnetwork/heimdall/pull/1082/commits/159a33bafa2f43c7cc2ae0f9d57b4257ef036436)

File: staking/README.md (hunk 1 above)
Comment (marcello33) [reply staking/README.md]: Yeah like that more, thanks.

File: staking/README.md (hunk 2 above)
Comment (marcello33) [reply staking/README.md:8]: I actually checked and apparently you need a `tab` (or two spaces) and not a double `*` to make it a sub-paragraph, sorry for the confusion (that works for headers like # and ##).
So in this case it'd be
```
* [How does one join the network as a validator](#how-does-one-join-the-network-as-a-validator)
  * [How to propose a MsgValidatorJoin transaction](#how-to-propose-a-msgvalidatorjoin-transaction)
```

Comment (Raneet10) [reply staking/README.md]: And now: [0ce7e54](https://github.com/maticnetwork/heimdall/pull/1082/commits/0ce7e545e376b2cc4cb0b4027731f326607523fa) ?

**Review by marcello33 — APPROVED** [review]

**Review by 0xKrishna — APPROVED** [review]

**Pull request merged**